# Servo Config for EB Robot
# Limits defined here are absolute limits (before something breaks).  Limits in code/Moveit should be smaller than these.

## NOTE: for reversed servo, swap min and max values! (min > max)

## SERVO SPEEDS USED:
# MX28:   55 RPM --> 5.8 rad/sec (2.5 NM Torque)
# MX64:   63 RPM --> 6.6 rad/sec (6.0 NM Torque)
# MX106:  35 RPM --> 3.7 rad/sec (8.4 NM Torque)

# Common reference points
#            Degrees   Ticks     Radians
# Min:      -180 deg       1    -3.14159
#            -90 deg    1024    -1.5708
#            -45 deg    1536    -0.785398
#            -20 deg    1844    -0.349066
#            -15 deg    1895    -0.261799
#            -10 deg            -0.174533
# Center:      0 deg    2048     0.0
#              1 deg             0.0174533
#             10 deg             0.174533
#             15 deg    2201     0.261799
#             20 deg    2252     0.349066
#             45 deg    2560     0.785398
#             90 deg    3072     1.5708
# Max:      +180 deg    4095     3.14159


## WARNING!  IF YOU ADD OR REMOVE SERVOS, CHANGE "max_motor_id" in "servos.launch"
# Note: IF USED, hip_lean_joints is a dual servo joint of left and right hip lean servos!

dynamixels: [

  'right_antenna_joint', 'left_antenna_joint', 
  'head_sidetilt_joint', 'head_pan_joint', 
  'head_tilt_joint',     'neck_raise_joint',

  'right_leg_hip_rotate_joint',   
  'right_leg_thigh_lift_joint',     
  'right_leg_knee_bend_joint',
  'right_leg_ankle_rotate_joint', 

  'left_leg_hip_rotate_joint',   
  'left_leg_thigh_lift_joint',     
  'left_leg_knee_bend_joint',
  'left_leg_ankle_rotate_joint'] 


controllers: [

  'right_antenna_joint', 'left_antenna_joint', 
  'head_sidetilt_joint', 'head_pan_joint', 
  'head_tilt_joint',     'neck_raise_joint',

  'right_leg_hip_rotate_joint',   
  'right_leg_thigh_lift_joint',     
  'right_leg_knee_bend_joint',
  'right_leg_ankle_rotate_joint', 

  'left_leg_hip_rotate_joint',   
  'left_leg_thigh_lift_joint',     
  'left_leg_knee_bend_joint',
  'left_leg_ankle_rotate_joint'] 



# --------------------------------------------------------------
# Head

right_antenna_joint: # Positive = Forward
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_antenna_joint
    joint_speed: 5.8
    motor:
        id: 1
        init: 2047
        max: 1024   # +90  REVERSED SERVO
        min: 3072   # -90  
        

left_antenna_joint: # Positive = Forward
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_antenna_joint
    joint_speed: 5.8
    motor:
        id: 2
        init: 2047
        min: 1024   # -90  
        max: 3072   # +90  


head_sidetilt_joint: # Positive = Tilt Right
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: head_sidetilt_joint
    joint_speed: 5.8
    motor:
        id: 3
        init: 2047
        min: 1535   # -45  
        max: 2560   # +45  
        
        
head_pan_joint: # Positive = Pan Right 
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: head_pan_joint
    joint_speed: 5.8
    motor:
        id: 4
        init: 2047
        min: 1024   # look right - limit to 90 deg
        max: 3072   # look left

head_tilt_joint: # Positive = Look Up
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: head_tilt_joint
    joint_speed: 5.8
    motor:
        id: 5
        init: 2047
        max: 1032  # look up REVERSED SERVO
        min: 3072   # -90 look down limit
        
neck_raise_joint: # Positive = Raise Head Up / Forward (but positive limit is just past zero)
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: neck_raise_joint
    joint_speed: 5.8
    motor:
        id: 6
        init: 2047  # straight up
        max: 1700   # >20 deg forward, hitting front stop REVERSED SERVO
        min: 3072   # -90
        


# --------------------------------------------------------------
# Right leg

right_leg_hip_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_leg_hip_rotate_joint
    joint_speed: 5.8
    motor:
        id: 8
        init: 2047
        min:  1895 #1946 15 deg normal, 10 deg for testing
        max:  2201 #2150


right_leg_thigh_lift_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_leg_thigh_lift_joint
    joint_speed: 6.6 # MX-64
    motor:
        id: 9
        init: 2047
        min: 1024   # -90  
        max: 3072   # +90  

right_leg_knee_bend_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_leg_knee_bend_joint
    joint_speed: 3.7 # 3.7 MX106
    motor:
        id: 10
        init: 2047  
        max: 383    # -145 deg SITTING POSITION!  REVERSED SERVO
        min: 3072   # +90  

right_leg_ankle_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: right_leg_ankle_rotate_joint
    joint_speed: 6.6 # MX-64
    motor:
        id: 11
        init: 2047 # TUNED!
        min: 1024   # -90
        max: 3072   # +90



# --------------------------------------------------------------
# left leg - most servos REVERSED

left_leg_hip_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_leg_hip_rotate_joint
    joint_speed: 5.8
    motor:
        id: 13
        init: 2047
        max:  1895 #1946 15 deg normal, 10 deg for testing
        min:  2201 #2150

left_leg_thigh_lift_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_leg_thigh_lift_joint
    joint_speed: 6.6
    motor:
        id: 14
        init: 2047
        max: 1024   # -90  REVERSED SERVO
        min: 3072   # +90  

left_leg_knee_bend_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_leg_knee_bend_joint
    joint_speed: 3.7 # 3.7 MX106
    motor:
        id: 15
        init: 2047
        min: 1024    # 90 degrees NORMAL SERVO
        max: 3697   # -145 deg SITTING POSITION! 

left_leg_ankle_rotate_joint:
    controller:
        package: dynamixel_controllers
        module: joint_position_controller
        type: JointPositionController
    joint_name: left_leg_ankle_rotate_joint
    joint_speed: 6.6 # MX-64
    motor:
        id: 16
        init: 2057 # TUNED!
        max: 1024   # -20 REVERSED SERVO
        min: 3072   # +20




 

